File: scripts/fix_csv_taxonomy.py
import csv
import sys

INPUT_FILE = 'attached_assets/2026_01_(1)_1767837875752.csv'
OUTPUT_FILE = 'attached_assets/corrected_2026_races.csv'

VALID_SPORTS = ['Running', 'Triathlon', 'Cycling', 'Fitness', 'Swimming', 'Custom']

VALID_SUBTYPES = {
    'Running': ['5K', '10K', 'Half Marathon', 'Marathon', 'Ultra Marathon', 'Trail Running', 'Cross Country', 'Custom Distance'],
    'Triathlon': ['Sprint', 'Olympic', 'Half Ironman', 'Ironman', 'Aquathlon', 'Duathlon', 'Custom Distance'],
    'Cycling': ['Criterium', 'Gran Fondo', 'Mountain Biking', 'Road Race', 'Custom Distance'],
    'Fitness': ['Spartan Race', 'HYROX', 'Obstacle Course', 'CrossFit', 'Bootcamp', 'Custom Distance'],
    'Swimming': ['Open Water Swim', 'Pool Competition', 'Custom Distance'],
    'Custom': ['Custom Event']
}

def fix_row(row):
    sport = row.get('sport', '').strip()
    sport_category = row.get('sport_category', '').strip()
    sport_subtype = row.get('sport_subtype', '').strip()
    
    fixed_sport = sport
    fixed_subtype = sport_subtype
    
    if sport == 'Spartan':
        fixed_sport = 'Fitness'
        if sport_subtype in ['Super', 'Beast', 'Open', 'Sprint', 'Trifecta', '']:
            fixed_subtype = 'Spartan Race'
    
    elif sport == 'Hyrox':
        fixed_sport = 'Fitness'
        if sport_subtype in ['Pro', '']:
            fixed_subtype = 'HYROX'
    
    elif sport == 'Running':
        if sport_subtype == 'Ultra':
            fixed_subtype = 'Ultra Marathon'
        elif sport_subtype == 'Trail':
            fixed_subtype = 'Trail Running'
        elif sport_subtype == 'Ultra Trail':
            fixed_subtype = 'Ultra Marathon'
        elif sport_subtype in ['20K', '6.7K']:
            fixed_subtype = 'Custom Distance'
        elif sport_subtype == '':
            if 'Trail' in sport_category:
                fixed_subtype = 'Trail Running'
            elif 'Road' in sport_category:
                if 'marathon' in row.get('name', '').lower():
                    fixed_subtype = 'Marathon'
                elif 'half' in row.get('name', '').lower():
                    fixed_subtype = 'Half Marathon'
                elif '10k' in row.get('name', '').lower() or '10 k' in row.get('name', '').lower():
                    fixed_subtype = '10K'
                elif '5k' in row.get('name', '').lower() or '5 k' in row.get('name', '').lower():
                    fixed_subtype = '5K'
                else:
                    fixed_subtype = 'Custom Distance'
            else:
                fixed_subtype = 'Custom Distance'
    
    elif sport == 'Triathlon':
        if sport_subtype == '':
            if 'Olympic' in sport_category:
                fixed_subtype = 'Olympic'
            elif 'Half Ironman' in sport_category or '70.3' in sport_category:
                fixed_subtype = 'Half Ironman'
            elif 'Full Ironman' in sport_category or 'Ironman' in sport_category:
                fixed_subtype = 'Ironman'
            elif 'Sprint' in sport_category:
                fixed_subtype = 'Sprint'
            else:
                fixed_subtype = 'Olympic'
        elif sport_subtype == 'Open':
            fixed_subtype = 'Olympic'
    
    if fixed_subtype not in VALID_SUBTYPES.get(fixed_sport, []):
        if fixed_sport in VALID_SUBTYPES:
            fixed_subtype = 'Custom Distance'
    
    row['sport'] = fixed_sport
    row['sport_category'] = ''
    row['sport_subtype'] = fixed_subtype
    
    return row

def main():
    rows_processed = 0
    rows_fixed = 0
    
    with open(INPUT_FILE, 'r', encoding='utf-8-sig') as infile:
        content = infile.read()
        if content.startswith('```'):
            lines = content.split('\n')
            lines = [l for l in lines if not l.startswith('```')]
            content = '\n'.join(lines)
        
        reader = csv.DictReader(content.splitlines())
        fieldnames = reader.fieldnames
        
        fixed_rows = []
        for row in reader:
            original = dict(row)
            fixed = fix_row(row)
            if original != fixed:
                rows_fixed += 1
            fixed_rows.append(fixed)
            rows_processed += 1
    
    with open(OUTPUT_FILE, 'w', newline='', encoding='utf-8') as outfile:
        writer = csv.DictWriter(outfile, fieldnames=fieldnames)
        writer.writeheader()
        writer.writerows(fixed_rows)
    
    print(f"Processed {rows_processed} rows")
    print(f"Fixed {rows_fixed} rows with taxonomy issues")
    print(f"Output saved to: {OUTPUT_FILE}")
    
    sport_counts = {}
    subtype_counts = {}
    for row in fixed_rows:
        sport = row['sport']
        subtype = row['sport_subtype']
        sport_counts[sport] = sport_counts.get(sport, 0) + 1
        subtype_counts[subtype] = subtype_counts.get(subtype, 0) + 1
    
    print("\n--- Sport Distribution ---")
    for sport, count in sorted(sport_counts.items(), key=lambda x: -x[1]):
        print(f"  {sport}: {count}")
    
    print("\n--- Subtype Distribution ---")
    for subtype, count in sorted(subtype_counts.items(), key=lambda x: -x[1]):
        print(f"  {subtype}: {count}")

if __name__ == '__main__':
    main()

File: scripts/fix_csv_taxonomy.js
const fs = require('fs');

const INPUT_FILE = 'attached_assets/2026_01_(1)_1767837875752.csv';
const OUTPUT_FILE = 'attached_assets/corrected_2026_races.csv';

const VALID_SUBTYPES = {
    'Running': ['5K', '10K', 'Half Marathon', 'Marathon', 'Ultra Marathon', 'Trail Running', 'Cross Country', 'Custom Distance'],
    'Triathlon': ['Sprint', 'Olympic', 'Half Ironman', 'Ironman', 'Aquathlon', 'Duathlon', 'Custom Distance'],
    'Cycling': ['Criterium', 'Gran Fondo', 'Mountain Biking', 'Road Race', 'Custom Distance'],
    'Fitness': ['Spartan Race', 'HYROX', 'Obstacle Course', 'CrossFit', 'Bootcamp', 'Custom Distance'],
    'Swimming': ['Open Water Swim', 'Pool Competition', 'Custom Distance'],
    'Custom': ['Custom Event']
};

function parseCSV(content) {
    const lines = content.split('\n').filter(l => l.trim() && !l.startsWith('```'));
    const headers = lines[0].split(',').map(h => h.trim());
    const rows = [];
    
    for (let i = 1; i < lines.length; i++) {
        const values = [];
        let current = '';
        let inQuotes = false;
        
        for (const char of lines[i]) {
            if (char === '"') {
                inQuotes = !inQuotes;
            } else if (char === ',' && !inQuotes) {
                values.push(current.trim());
                current = '';
            } else {
                current += char;
            }
        }
        values.push(current.trim());
        
        if (values.length >= headers.length) {
            const row = {};
            headers.forEach((h, idx) => row[h] = values[idx] || '');
            rows.push(row);
        }
    }
    return { headers, rows };
}

function fixRow(row) {
    let sport = (row.sport || '').trim();
    let sportCategory = (row.sport_category || '').trim();
    let sportSubtype = (row.sport_subtype || '').trim();
    const name = (row.name || '').toLowerCase();
    
    if (sport === 'Spartan') {
        sport = 'Fitness';
        sportSubtype = 'Spartan Race';
    }
    else if (sport === 'Hyrox') {
        sport = 'Fitness';
        sportSubtype = 'HYROX';
    }
    else if (sport === 'Running') {
        if (sportSubtype === 'Ultra') {
            sportSubtype = 'Ultra Marathon';
        } else if (sportSubtype === 'Trail') {
            sportSubtype = 'Trail Running';
        } else if (sportSubtype === 'Ultra Trail') {
            sportSubtype = 'Ultra Marathon';
        } else if (['20K', '6.7K'].includes(sportSubtype)) {
            sportSubtype = 'Custom Distance';
        } else if (!sportSubtype) {
            if (sportCategory.includes('Trail')) {
                sportSubtype = 'Trail Running';
            } else if (sportCategory.includes('Road')) {
                if (name.includes('marathon') && !name.includes('half')) {
                    sportSubtype = 'Marathon';
                } else if (name.includes('half marathon') || name.includes('half-marathon')) {
                    sportSubtype = 'Half Marathon';
                } else if (name.includes('10k') || name.includes('10 k')) {
                    sportSubtype = '10K';
                } else if (name.includes('5k') || name.includes('5 k')) {
                    sportSubtype = '5K';
                } else {
                    sportSubtype = 'Custom Distance';
                }
            } else {
                sportSubtype = 'Custom Distance';
            }
        }
    }
    else if (sport === 'Triathlon') {
        if (!sportSubtype || sportSubtype === 'Open') {
            if (sportCategory.includes('Olympic')) {
                sportSubtype = 'Olympic';
            } else if (sportCategory.includes('Half Ironman') || sportCategory.includes('70.3')) {
                sportSubtype = 'Half Ironman';
            } else if (sportCategory.includes('Full Ironman') || (sportCategory.includes('Ironman') && !sportCategory.includes('Half'))) {
                sportSubtype = 'Ironman';
            } else if (sportCategory.includes('Sprint')) {
                sportSubtype = 'Sprint';
            } else {
                sportSubtype = 'Olympic';
            }
        }
    }
    
    const validForSport = VALID_SUBTYPES[sport] || [];
    if (!validForSport.includes(sportSubtype)) {
        sportSubtype = 'Custom Distance';
    }
    
    row.sport = sport;
    row.sport_category = '';
    row.sport_subtype = sportSubtype;
    
    return row;
}

function toCSV(headers, rows) {
    const lines = [headers.join(',')];
    for (const row of rows) {
        const values = headers.map(h => {
            let val = row[h] || '';
            if (val.includes(',') || val.includes('"')) {
                val = '"' + val.replace(/"/g, '""') + '"';
            }
            return val;
        });
        lines.push(values.join(','));
    }
    return lines.join('\n');
}

function main() {
    let content = fs.readFileSync(INPUT_FILE, 'utf-8');
    if (content.charCodeAt(0) === 0xFEFF) {
        content = content.slice(1);
    }
    
    const { headers, rows } = parseCSV(content);
    console.log(`Loaded ${rows.length} rows`);
    
    let fixedCount = 0;
    const fixedRows = rows.map(row => {
        const original = JSON.stringify(row);
        const fixed = fixRow({ ...row });
        if (JSON.stringify(fixed) !== original) fixedCount++;
        return fixed;
    });
    
    const output = toCSV(headers, fixedRows);
    fs.writeFileSync(OUTPUT_FILE, output, 'utf-8');
    
    console.log(`Fixed ${fixedCount} rows with taxonomy issues`);
    console.log(`Output saved to: ${OUTPUT_FILE}`);
    
    const sportCounts = {};
    const subtypeCounts = {};
    fixedRows.forEach(row => {
        sportCounts[row.sport] = (sportCounts[row.sport] || 0) + 1;
        subtypeCounts[row.sport_subtype] = (subtypeCounts[row.sport_subtype] || 0) + 1;
    });
    
    console.log('\n--- Sport Distribution ---');
    Object.entries(sportCounts)
        .sort((a, b) => b[1] - a[1])
        .forEach(([sport, count]) => console.log(`  ${sport}: ${count}`));
    
    console.log('\n--- Subtype Distribution ---');
    Object.entries(subtypeCounts)
        .sort((a, b) => b[1] - a[1])
        .forEach(([subtype, count]) => console.log(`  ${subtype}: ${count}`));
}

main();
